jwst/pipeline/calwebb_image2.py

```diff
@@ -37,48 +40,68 @@ def process(self, input):
 
         return input
 
-
-    def setup_output(self, input):
-
-        # This routine doesn't actually save the final result to a file,
-        # but just sets up the value of self.output_file appropriately.
-        # The final data model is passed back up to the caller, which can be
-        # either an interactive session or a command-line instance of stpipe.
-        # If it's an interactive session, the data model is simply returned to
-        # the user without saving to a file. If it's a command-line instance
-        # of stpipe, stpipe will save the data model to a file using the name
-        # given in self.output_file.
-
-        # first determine the proper file name suffix to use later
-        if isinstance(input, datamodels.CubeModel):
-            suffix = 'calints'
+    # Process each exposure
+    def process_exposure_product(
+            self,
+            exp_product,
+            pool_name=' ',
+            asn_file=' '
+    ):
+        """Process an exposure found in the association product
+
+        Parameters
+        ---------
+        exp_product: dict
+            A Level2b association product.
+        """
+        # Find all the member types in the product
+        members_by_type = defaultdict(list)
+        for member in exp_product['members']:
+            members_by_type[member['exptype']].append(member['expname'])
+
+        # Get the science member. Technically there should only be
+        # one. We'll just get the first one found.
+        science = members_by_type['SCIENCE']
+        if len(science) != 1:
+            log.warn(
+                'Wrong number of science files found in {}'.format(
+                    exp_product['name']
+                )
+            )
+        science = science[0]
+
+        self.log.info('Working on input %s ...', science)
+        if isinstance(science, datamodels.DataModel):
+            input = science
         else:
-            suffix = 'cal'
-
-        # Has an output file name already been set?
-        if self.output_file is not None:
+            input = datamodels.open(science)
 
-            # Check to see if the output_file name is the default set by
-            # stpipe for command-line processing
-            root, ext = os.path.splitext(self.output_file)
-            if root[root.rfind('_') + 1:] == 'Image2Pipeline':
+        # Record ASN pool and table names in output
+        input.meta.asn.pool_name = pool_name
+        input.meta.asn.table_name = asn_file
 
-                # Remove the step name that stpipe appended to the file name,
-                # as well as the original suffix on the input file name,
-                # and create a new name with the appropriate output suffix
-                root = root[:root.rfind('_')]
-                self.output_file = root[:root.rfind('_') + 1] + suffix + ext
+        # work on slope images
+        input = self.assign_wcs(input)
+        input = self.flat_field(input)
```

Comment:
Should go ahead and take the opportunity to add a call to background_subtract (after assign_wcs) if there are background image(s) defined in the ASN, just as is done in calwebb_spec2. This something the JCCWG has requested for future expansion. Probably won't encounter any instances using background any time soon, but at least the code is ready to support it.